src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs

```diff
@@ -584,5 +584,1027 @@ void M()
+        public async Task TestObjectCreationAddTypeParameter_GeneralizeGenericArgument1()
+        {
+            var code =
+@"
+    class C1<T>
+    {
+        C1() { }
+        void M2()
+        {
+            var c = new [|C1<bool>|](true);
+        }
+    }
+";
+            var fix0 =
+@"
+    class C1<T>
+    {
+        C1(T v) { }
+        void M2()
+        {
+            var c = new C1<bool>(true);
+        }
+    }
+";
+            await TestInRegularAndScriptAsync(code, fix0);
+        }
+
+        [WorkItem(21446, "https://github.com/dotnet/roslyn/issues/21446")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddParameter)]
+        public async Task TestObjectCreationAddTypeParameter_GeneralizeGenericArgument2()
+        {
+            var code =
+@"
+    class C1<T>
+    {
+        C1() { }
+
+        void M1<U>() { }
+        void M2()
+        {
+            var c = new C1<bool>();
+            c.[|M1<int, bool>|]();
+        }
+    }
+";
+            var fix0 =
+@"
+    class C1<T>
+    {
+        C1() { }
+
+        void M1<U, T1>() { }
+        void M2()
+        {
+            var c = new C1<bool>();
+            c.M1<int, bool>();
+        }
+    }
+";
+            await TestInRegularAndScriptAsync(code, fix0);
+        }
+
+        [WorkItem(21446, "https://github.com/dotnet/roslyn/issues/21446")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddParameter)]
+        public async Task TestObjectCreationAddTypeParameter_GeneralizeGenericArgument3()
+        {
+            var code =
+@"
+    class C1<T>
+    {
+        C1() { }
+
+        void M1<U>() { }
+        void M2()
+        {
+            var c = new C1<bool>();
+            c.[|M1<int, string>|]();
+        }
+    }
+";
+            var fix0 =
+@"
+    class C1<T>
+    {
+        C1() { }
+
+        void M1<U, T1>() { }
+        void M2()
+        {
+            var c = new C1<bool>();
+            c.M1<int, string>();
+        }
+    }
+";
+            await TestInRegularAndScriptAsync(code, fix0);
+        }
+
+        [WorkItem(21446, "https://github.com/dotnet/roslyn/issues/21446")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddParameter)]
+        public async Task TestObjectCreationAddTypeParameter_GeneralizeGenericArgument4()
+        {
+            var code =
+@"
+    class C1<T> where T: struct
+    {
+        C1() { }
+
+        void M1<U>() where U : struct
+        {
+        }
+        void M2()
+        {
+            var c = new C1<bool>();
+            c.[|M1<int, string>|]();
+        }
+    }
+";
+            var fix0 =
+@"
+    class C1<T> where T: struct
+    {
+        C1() { }
+
+        void M1<U, T1>() where U : struct
+        {
+        }
+        void M2()
+        {
+            var c = new C1<bool>();
+            c.M1<int, string>();
+        }
+    }
+";
+            await TestInRegularAndScriptAsync(code, fix0);
```

Comment:
one thing i don't think i saw was people inserting arguments into the middle of a working list and having that generate the parameter in the right location. I'd like to see this with interesting combinations of:

1. generics.
2. params methods.
3. named parameters.